packages/editor/src/components/post-template/hooks.js

```diff
@@ -55,27 +55,41 @@ function useTemplates( postType ) {
 	// To do: create a new selector to checks if templates exist at all instead
 	// of and unbound request. In the modal, the user templates should be
 	// paginated and we should not make an unbound request.
-	const { staticTemplates, templates } = useSelect(
+	return useSelect(
 		( select ) => {
-			return {
-				staticTemplates: select( coreStore ).getEntityRecords(
-					'postType',
-					'wp_registered_template',
-					{ per_page: -1, post_type: postType }
-				),
-				templates: select( coreStore ).getEntityRecords(
-					'postType',
-					'wp_template',
-					{ per_page: -1, post_type: postType }
+			const currentTheme = select( coreStore ).getCurrentTheme();
+			const registeredTemplates = select( coreStore ).getEntityRecords(
+				'postType',
+				'wp_registered_template',
+				{ per_page: -1, post_type: postType }
+			);
+			const userTemplates = select( coreStore ).getEntityRecords(
+				'postType',
+				'wp_template',
+				{ per_page: -1 }
+			);
+			if ( ! currentTheme || ! registeredTemplates || ! userTemplates ) {
+				return [];
+			}
+
+			const defaultTemplateTypes = currentTheme.default_template_types;
+			return [
+				...registeredTemplates,
+				...userTemplates.filter(
+					( template ) =>
+						// Only give "custom" templates as an option, which
+						// means the is_wp_suggestion meta field is not set and
+						// the slug is not found in the default template types.
+						// https://github.com/WordPress/wordpress-develop/blob/97382397b2bd7c85aef6d4cd1c10bafd397957fc/src/wp-includes/block-template-utils.php#L858-L867
+						! template.meta.is_wp_suggestion &&
+						! defaultTemplateTypes.find(
+							( type ) => type.slug === template.slug
+						)
 				),
-			};
+			];
```

Comment:
This is a practice we're trying to discourage and use rarely. It's probably better to do this merging and filtering in `useMemo`, no?

Reply:
I know it's not the case here, but it's easy to "break" shallow equality checks if you're not careful and manipulate objects/arrays inside the `useSelect` callback, IMO.

We even introduced a dev warning to catch those early and help third-party developers.

Reply:
Hm, but the same can be said of returning an object breaking shallow equality if you nest them. I get your argument, but I think there are valid cases like being able avoid passing values to the render function and avoid re-renders due to value changes. Not super applicable here, but I like avoiding the `useMemo` call if it's possible like this.